[GPU Process] Focus outline is off on <select> on wpt.fyi/interop-2023

pxlcoder opened the pull request:
#### be4588f1fb254d36603c59d981fa0dfc44143020
<pre>
[GPU Process] Focus outline is off on &lt;select&gt; on wpt.fyi/interop-2023
<a href="https://bugs.webkit.org/show_bug.cgi?id=254472">https://bugs.webkit.org/show_bug.cgi?id=254472</a>
rdar://107165828

Reviewed by Simon Fraser.

WebKit uses a flipped context to draw `NSCell`s. However, when cells are drawn
without a control view (as is the case with GPU Process enabled), AppKit does
not check the flippedness of the context, and unconditionally assumes an
unflipped context.

A workaround for this behavior was applied for cell drawing in 261085@main, but
was not extended to cell focus ring drawing. Note that support for viewless cell
focus ring drawing was only added by AppKit in rdar://105249508.

To fix, bring the same logic from 261085@main over to cell focus ring drawing.

* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::performDrawingWithUnflippedContext):

Add a helper method to draw with an unflipped context.

(WebCore::drawSliderCell):
(WebCore::drawViewlessCell):
(WebCore::drawViewlessCellFocusRing):
(WebCore::drawCellFocusRingInView):
(WebCore::ControlMac::drawCellFocusRingInternal):

Canonical link: <a href="https://commits.webkit.org/262183@main">https://commits.webkit.org/262183@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/038bc2ffd3e295ef6664f088423d4a214005fb3f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/661 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/680 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/708 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/834 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/579 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/652 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/729 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/765 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/818 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/668 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/630 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/628 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/812 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/691 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/614 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/649 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/592 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/636 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/624 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/614 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/577 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/629 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/197 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/634 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (approved):
Is there a reliable way to test this?

pxlcoder committed:
[GPU Process] Focus outline is off on <select> on wpt.fyi/interop-2023
https://bugs.webkit.org/show_bug.cgi?id=254472
rdar://107165828

Reviewed by Simon Fraser.

WebKit uses a flipped context to draw `NSCell`s. However, when cells are drawn
without a control view (as is the case with GPU Process enabled), AppKit does
not check the flippedness of the context, and unconditionally assumes an
unflipped context.

A workaround for this behavior was applied for cell drawing in 261085@main, but
was not extended to cell focus ring drawing. Note that support for viewless cell
focus ring drawing was only added by AppKit in rdar://105249508.

To fix, bring the same logic from 261085@main over to cell focus ring drawing.

* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::performDrawingWithUnflippedContext):

Add a helper method to draw with an unflipped context.

(WebCore::drawSliderCell):
(WebCore::drawViewlessCell):
(WebCore::drawViewlessCellFocusRing):
(WebCore::drawCellFocusRingInView):
(WebCore::ControlMac::drawCellFocusRingInternal):

Canonical link: https://commits.webkit.org/262183@main
